kubernetes: Integrate k8s common labels

mig4 opened the pull request:
Towards https://github.com/giantswarm/giantswarm/issues/8632

Include common labels advocated in Kubernetes docs, there's some
overlap with our internal ones and after discussing this the consensus
was that it would be better to avoid that and standardise on the well
defined set from Kubernetes/Helm upstream.

Those labels allow us to describe the source and instance of an app so
it's easy to see which installation an object is part of.

Inline description of the main labels specifying what are the expected
values and that they apply to all objects composing an app/operator
they refer to.

Also add this to description of related custom `.../branch` and
`.../commit` labels.

mig4 committed:
kubernetes: Integrate k8s common labels

Towards https://github.com/giantswarm/giantswarm/issues/8632

Include common labels advocated in Kubernetes docs, there's some
overlap with our internal ones and after discussing this the consensus
was that it would be better to avoid that and standardise on the well
defined set from Kubernetes/Helm upstream.

Those labels allow us to describe the source and instance of an app so
it's easy to see which installation an object is part of.

Inline description of the main labels specifying what are the expected
values and that they apply to all objects composing an app/operator
they refer to.

Also add this to description of related custom `.../branch` and
`.../commit` labels.

mig4 committed:
Remove `app` and `version` labels

Now replaced with upstream Kubernetes/Helm recommended labels
`app.kubernetes.io/name` and `app.kubernetes.io/version`.

mig4 committed:
List labels applied in Helm charts

Add a full list of app labels applied in Helm charts so it's easier
to see without also having to look at external docs.

kopiczko commented on kubernetes/annotations_and_labels.md:
```diff
@@ -51,16 +47,38 @@ This page defines common annotations and labels we set in Kubernetes objects.
+- `app.kubernetes.io/name` - the name of the application; should be applied to
+  every Kubernetes resource associated with that application, i.e. all objects
+  in a chart that installs it.
+- `app.kubernetes.io/instance` - value should be set to `{{ .Release.Name }}`
```
It's somehow weird that we say what should be set here but not in the other labels.

mig4 replied:
Well, the Helm docs say `"This should be the {{ .Release.Name }}. "` so I thought it'd be best to include it here.

mig4 replied:
I added example of using template helper at the bottom.

kopiczko commented on kubernetes/annotations_and_labels.md:
```diff
@@ -51,16 +47,38 @@ This page defines common annotations and labels we set in Kubernetes objects.
+  [here][helm-labels] and [here][k8s-common-labels]) - are informational,
+  applied to all objects in a chart that installs an app or operator and so
+  they indicate the source and instance of that app/operator; value of
+  `.../version` should be the app/operator version as defined in `project.go`
```
This `.../version` shortcut is confusing IMO.

mig4 replied:
Fair enough, I'll get rid of them.

mig4 replied:
Done

kopiczko commented on kubernetes/annotations_and_labels.md:
```diff
@@ -51,16 +47,38 @@ This page defines common annotations and labels we set in Kubernetes objects.
+  selector, e.g.
+  `app.kubernetes.io/name=kvm-operator,app.kubernetes.io/instance=kvm-operator-1.0.0`.
+- `app.kubernetes.io/version` and other common labels (see
+  [here][helm-labels] and [here][k8s-common-labels]) - are informational,
```
I would skip the other common labels part. This page is about labels we use and how we do this. If we want to use any of the common upstream labels it should be listed here.

mig4 replied:
Sure, makes sense.

mig4 replied:
Done

kopiczko commented on kubernetes/annotations_and_labels.md:
```diff
@@ -51,16 +47,38 @@ This page defines common annotations and labels we set in Kubernetes objects.
+  [Helm chart labels best practices][helm-labels]
+- `app.kubernetes.io/name` - the name of the application; should be applied to
+  every Kubernetes resource associated with that application, i.e. all objects
+  in a chart that installs it.
```
It would be cool to add some example here.

mig4 replied:
Done

kopiczko commented on kubernetes/annotations_and_labels.md:
```diff
@@ -161,6 +179,20 @@ This page defines common annotations and labels we set in Kubernetes objects.
 
 #### Tenant Cluster
 
+### App Labels
```
Someone decided to skip one level in the following paragraphs so I think you want to do `##`.

```suggestion
## Chart Templates Labels
```

mig4 replied:
Do you mean that all the sub-headings of _"Common Labels"_ should be one level up? E.g. that `### Labels Set In Custom Resources` should actually be `## Labels Set In Custom Resources`?

kopiczko replied:
Sorry my mind was in a dark place apparently. `###` is good.

kopiczko replied:
Updated my suggestion.

mig4 replied:
So how about something more explicit like _Labels Set On Objects Installed By Charts_?

mig4 replied:
Went with that, let me know if that works.

mig4 committed:
Group informational labels, improve descriptions

* group informational labels so there's no need to repeat a good part
  of their description
* don't use `.../version` shortcut
* skip _other common labels_ from the list, only mention the ones we
  use internally there

mig4 committed:
Add example of `app.kubernetes.io/name` label

mig4 committed:
Revise title of app labels section

mig4 committed:
Revise label indicating version of reconciling operator

* remove it from list applied to objects installed by charts as it
  makes no sense (operators don't reconcile objects they're composed of)
* highlight that it applies to objects the given operator reconciles

mig4 committed:
Explain maintaining labels using helper template

This allows maintaining them in one place and using in multiple places
without repetition. It is based on approach in default chart template
in upstream Helm.

Also add a collapsed by default example of how this is rendered.

mig4 committed:
Fix wording and formatting

mig4 committed:
Sort labels alphabetically

mig4 committed:
Restore `app` label

kopiczko commented on kubernetes/annotations_and_labels.md:
```diff
@@ -16,9 +16,10 @@ This page defines common annotations and labels we set in Kubernetes objects.
+  to every Kubernetes resource associated with an application. E.g.
+  `app=kvm-operator`. This is softly deprecated and should be replaced with
+  `app.kubernetes.io/name`, it's only kept to avoid breaking existing
+  workflows.
```
I'd put it like:

```
- `app` - should be set to the same value as `app.kubernetes.io/name`. It's only kept
  to avoid breaking existing workflows.
```

mig4 replied:
Done

kopiczko commented on kubernetes/annotations_and_labels.md:
```diff
@@ -160,6 +182,156 @@ This page defines common annotations and labels we set in Kubernetes objects.
+  (adjust indent as required)
+- `aws-operator.selectorLabels` - defines labels to be used in selectors;
+  usage: `{{- include "aws-operator.selectorLabels" . | nindent 6 }}`
+  (adjust indent as required)
```
```suggestion
  usage: `{{- include "aws-operator.selectorLabels" . | nindent INDENT }}`
  (set INDENT to required integer)
```

mig4 replied:
Done

kopiczko commented on kubernetes/annotations_and_labels.md:
```diff
@@ -160,6 +182,156 @@ This page defines common annotations and labels we set in Kubernetes objects.
+  to 63 characters and with `+` signs replaced with `-`
+- `aws-operator.labels` - defines all the labels described above, including the
+  selector labels; usage: `{{- include "aws-operator.labels" . | nindent 4 }}`
+  (adjust indent as required)
```
To avoid copy/paste errors.

```suggestion
  selector labels; usage: `{{- include "aws-operator.labels" . | nindent INDENT }}`
  (set INDENT to required integer)
```

mig4 replied:
Done

kopiczko reviewed (commented):
This is really nice. LGTM thanks <3

kopiczko reviewed (approved):
Forgot to approve!

mig4 committed:
Simplify description of `app` label

mig4 committed:
Be a bit less explicit with label snippet examples

To avoid copy/paste errors but improve explanation.

mig4 committed:
Remove lang from code block

Highlighting doesn't work for go-template/mustache anyway.

mig4 committed:
Move selector labels to the top

Co-Authored-By: Paweł Kopiczko <pawel@giantswarm.io>

mig4 commented:
We talked about this in team standup today, and I'm going to go with this in a bit and will write a post to #news-dev to make everyone aware.

wtty-fool commented on kubernetes/annotations_and_labels.md:
```diff
@@ -41,25 +39,46 @@ This page defines common annotations and labels we set in Kubernetes objects.
+  `app.kubernetes.io/name=kvm-operator,app.kubernetes.io/instance=kvm-operator-1.0.0`.
+- informational labels, applied to all objects in a chart that installs an app
+  or operator and so they indicate its source and instance:
+  - `app.kubernetes.io/version` - value should be the app/operator version as
```
Just a thought: since https://github.com/giantswarm/gatekeeper-control-plane-rules/pull/20, `app.kubernetes.io/version` is used for gatekeeper uniqueness checks [(see here)](https://github.com/giantswarm/gatekeeper-control-plane-rules/blob/master/helm/gatekeeper-control-plane-rules/templates/uniqueoperatorversion.yaml#L16).

It would be helpful to include this information here, somewhere.

mig4 replied:
Em, that's not right is it? I mean the operators still use the `$OPERATOR.giantswarm.io/version` label for marking objects they reconcile, e.g.:

- https://github.com/giantswarm/aws-operator/blob/master/service/controller/cluster_resource_set.go#L719
- https://github.com/giantswarm/aws-operator/blob/ad94f07f4a83ff2cb31b786deddf9ce64af92197/service/controller/key/common.go#L122
- https://github.com/giantswarm/aws-operator/blob/ad94f07f4a83ff2cb31b786deddf9ce64af92197/pkg/label/label.go#L18

I missed this change in GK rules, why was that done?

kopiczko replied:
GK rules are to make sure we don't install two deployments with the same `app.kubernetes.io/name` and `app.kubernetes.io/version`.

BTW @wtty-fool we should also rename `app` to `app.kubernetes.io/name` here https://github.com/giantswarm/gatekeeper-control-plane-rules/pull/20/files#diff-d639fe45fa938f521566f27263b98d82R17

mig4 replied:
Yeah, oh right so this is because we removed `$OPERATOR.giantswarm.io/version` from the operator's own `Deployment` object. Meaning it now compares the internal project version which is exposed as `app.kubernetes.io/version` label on its own components, to `$OPERATOR.giantswarm.io/version` set on reconciled objects. OK, makes sense, I'll update descriptions.

mig4 replied:
Done. I'm going to leave this until tomorrow so in case the wording can be improved let me know.

mig4 committed:
Explain which labels are used for uniqueness checks

kopiczko reviewed (approved):
I like the changes.

pipo02mix commented on kubernetes/annotations_and_labels.md:
```diff
@@ -15,10 +15,8 @@ This page defines common annotations and labels we set in Kubernetes objects.
-  label should also be used as replicas selector. E.g. `app=kvm-operator,version=1.0.0`. Exceptions can
-  be made to accomodate for adherence to existing selectors upstream.
+- `app` - should be set to the same value as `app.kubernetes.io/name`. It's
+  deprecated and only kept to avoid breaking existing workflows.
```
We need to announce in the releases we do in order to give customers to adapt automation

mig4 replied:
Sure, I'll mention that we need to do that in the ticket.

pipo02mix reviewed (approved):
this is awesome

kubasobon reviewed (approved):
Thanks

tomahawk28 reviewed (approved):
LGTM
